src/httpauth/digest.c

```diff
@@ -591,3 +613,405 @@ int httpauth_digest_chall_request_full(struct httpauth_digest_chall_req **preq,
+	}
+
+	if (err)
+		goto out;
+
+	resp->hash_function(mb->buf, mb->end, hash1);
+	mbuf_rewind(mb);
+
+	if (str_str(resp->algorithm, "-sess")) {
+		err = mbuf_printf(mb, "%w:%s:%s",
+			hash1, resp->hash_length, resp->nonce, resp->cnonce);
+		if (err)
+			goto out;
+
+		resp->hash_function(mb->buf, mb->end, hash1);
+		mbuf_rewind(mb);
+	}
+
+	/* DIGEST */
+	if (str_isset(resp->qop)) {
+		err = mbuf_printf(mb, "%w:%s:%s:%s:%s:%w",
+			hash1, resp->hash_length, resp->nonce, resp->nc,
+			resp->cnonce, resp->qop, hash2, resp->hash_length);
+	}
+	else {
+		err = mbuf_printf(mb, "%w:%s:%w", hash1, resp->hash_length,
+			resp->nonce, hash2, resp->hash_length);
+	}
+
+	if (err)
+		goto out;
+
+	resp->hash_function(mb->buf, mb->end, hash1);
+	n = re_snprintf(resp->response, hashstringl, "%w",
+		hash1, resp->hash_length);
+	if (n == -1 || n != (int)hashstringl - 1)
+		err = ERANGE;
+
+out:
+	mem_deref(mb);
+	mem_deref(hash1);
+	mem_deref(hash2);
+
+	return err;
+}
+
+
+/**
+ * Prints / encodes an HTTP digest response
+ *
+ * @param pf   Re_printf object
+ * @param resp Response to print
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int httpauth_digest_response_print(struct re_printf *pf,
+	const struct httpauth_digest_enc_resp *resp)
+{
+	int err = 0;
+
+	if (!resp)
+		return EINVAL;
+
+	/* historical reason quoted strings:   */
+	/*   username, realm, nonce, uri,      */
+	/*   response, cnonce, opaque          */
+	/* historical reason unquoted strings: */
+	/*   qop, algorithm, nc                */
+	err = re_hprintf(pf, "Digest realm=\"%s\","
+		" nonce=\"%s\", username=\"%s\", uri=\"%s\","
+		" response=\"%s\"",
+		resp->realm, resp->nonce, resp->username,
+		resp->uri, resp->response);
+
+	if (str_isset(resp->opaque))
+		err |= re_hprintf(pf, ", opaque=\"%s\"", resp->opaque);
+	if (str_isset(resp->algorithm))
+		err |= re_hprintf(pf, ", algorithm=%s", resp->algorithm);
+	if (str_isset(resp->qop))
+		err |= re_hprintf(pf, ", qop=%s, cnonce=\"%s\", nc=\"%s\"",
+			resp->qop, resp->cnonce, resp->nc);
+
+	if (resp->userhash)
+		err |= re_hprintf(pf, ", userhash=true");
+	if (str_isset(resp->charset))
+		err |= re_hprintf(pf, ", charset=\"%s\"", resp->charset);
+
+	return err;
+}
+
+
+/**
+ * Set cnonce and nc and recalculate the response value.
+ * This function should be used only for unit tests
+ *
+ * @param resp       Httpauth_new_digest_response object pointer
+ * @param chall      Received and decoded digest challenge
+ * @param method     Used method
+ * @param user       Username
+ * @param passwd     User password
+ * @param entitybody Entitybody if qop=auth-int
+ * @param cnonce     Cnonce
+ * @param nc_        Nonce counter
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int httpauth_digest_response_set_cnonce(struct httpauth_digest_enc_resp *resp,
+	const struct httpauth_digest_chall *chall, const struct pl *method,
+	const char *user,	const char *passwd, const char *entitybody,
+	const uint32_t cnonce, const uint32_t nc_)
```

Comment:
remove trailing underscore from "nc_"

const is not needed for params passed by value

Reply:
I didn't want to write out the `nonce_count` and `nc` is already an static internal variable. Therefore the underscore.
I will write it out and remove the const.